File: internal/compliance/classifier.go
// Package compliance provides cross-language compliance helpers for Phase 84.
// The SignalClassifier maps fired RiskSignal names to human-readable attack
// categories and is the Go counterpart of management/compliance/classifier.py.
// Both implementations must produce identical output for the same input — see
// the cross-language parity test in Makefile target test-phase-84-classifier-parity.
package compliance

import (
	"sort"
)

// CategoryEntry maps one signal name to its category and weight.
type CategoryEntry struct {
	Category string
	Weight   int
}

// DefaultSignalCategories mirrors DEFAULT_SIGNAL_CATEGORIES in classifier.py.
// Any change here MUST be mirrored in the Python file and vice versa.
var DefaultSignalCategories = map[string]CategoryEntry{
	"spamhaus_drop":        {Category: "known_malicious_network",   Weight: 100},
	"spamhaus_edrop":       {Category: "known_malicious_network",   Weight: 100},
	"tor_exit":             {Category: "tor_exit_node",             Weight: 95},
	"beaconing_detected":   {Category: "c2_beaconing",              Weight: 90},
	"ja4_blacklist":        {Category: "malicious_tls_fingerprint", Weight: 85},
	"abuseipdb_score_high": {Category: "reported_abuse",            Weight: 70},
	"tls_version_old":      {Category: "obsolete_tls",              Weight: 60},
	"sni_missing":          {Category: "automation_tool",           Weight: 55},
	"sni_ip_literal":       {Category: "automation_tool",           Weight: 55},
	"datacenter":           {Category: "datacenter_scanner",        Weight: 50},
	"asn_datacenter":       {Category: "datacenter_scanner",        Weight: 50},
	"country_blacklist":    {Category: "geo_blocked",               Weight: 40},
}

// FallbackCategory is returned when no signals match.
const FallbackCategory = "high_risk_score"

// SignalClassifier maps sets of fired RiskSignal names to attack categories.
type SignalClassifier struct {
	categories map[string]CategoryEntry
}

// NewSignalClassifier creates a classifier with the default signal categories.
func NewSignalClassifier() *SignalClassifier {
	cats := make(map[string]CategoryEntry, len(DefaultSignalCategories))
	for k, v := range DefaultSignalCategories {
		cats[k] = v
	}
	return &SignalClassifier{categories: cats}
}

// NewSignalClassifierWithOverrides creates a classifier where entries in
// overrides are merged over the defaults.  Unknown keys in overrides are
// added; known keys in overrides replace the default entry.
func NewSignalClassifierWithOverrides(overrides map[string]CategoryEntry) *SignalClassifier {
	clf := NewSignalClassifier()
	for k, v := range overrides {
		clf.categories[k] = v
	}
	return clf
}

// Categories returns a copy of the active signal→CategoryEntry mapping.
func (c *SignalClassifier) Categories() map[string]CategoryEntry {
	out := make(map[string]CategoryEntry, len(c.categories))
	for k, v := range c.categories {
		out[k] = v
	}
	return out
}

// Classify returns the attack category for the given set of fired signal names.
//
// Resolution rules:
//   - The entry with the highest weight wins.
//   - Ties are broken by alphabetical order of the category name (ascending).
//   - Unknown signal names are silently ignored.
//   - Empty or all-unknown signals return FallbackCategory.
func (c *SignalClassifier) Classify(signals []string) string {
	bestWeight := -1
	bestCategory := FallbackCategory

	for _, signal := range signals {
		entry, ok := c.categories[signal]
		if !ok {
			continue
		}
		if entry.Weight > bestWeight ||
			(entry.Weight == bestWeight && entry.Category < bestCategory) {
			bestWeight = entry.Weight
			bestCategory = entry.Category
		}
	}
	return bestCategory
}

// ClassifyEvent adds a "category" field to a copy of the event map.
// The original map is not modified.
func (c *SignalClassifier) ClassifyEvent(event map[string]interface{}) map[string]interface{} {
	out := make(map[string]interface{}, len(event)+1)
	for k, v := range event {
		out[k] = v
	}
	var signals []string
	if raw, ok := event["signals"]; ok {
		switch s := raw.(type) {
		case []string:
			signals = s
		case []interface{}:
			for _, item := range s {
				if str, ok := item.(string); ok {
					signals = append(signals, str)
				}
			}
		}
	}
	out["category"] = c.Classify(signals)
	return out
}

// ClassifyBatch returns a new slice of event maps with a "category" field added.
func (c *SignalClassifier) ClassifyBatch(events []map[string]interface{}) []map[string]interface{} {
	out := make([]map[string]interface{}, len(events))
	for i, e := range events {
		out[i] = c.ClassifyEvent(e)
	}
	return out
}

// AllCategories returns a sorted slice of distinct category names in the mapping.
func (c *SignalClassifier) AllCategories() []string {
	seen := make(map[string]struct{})
	for _, entry := range c.categories {
		seen[entry.Category] = struct{}{}
	}
	result := make([]string, 0, len(seen))
	for cat := range seen {
		result = append(result, cat)
	}
	sort.Strings(result)
	return result
}

File: management/compliance/classifier.py
"""Signal classifier — maps fired RiskSignal names to attack categories.

Used by PciDssPackBuilder to populate artefact 03_attack_classification.csv
and by the report renderer to build the "Value Delivered" section.

Resolution rules
----------------
- Highest weight among all fired (and known) signals wins.
- Ties are broken alphabetically by *category name* (deterministic).
- Signals not present in the mapping are silently ignored.
- An empty/all-unknown signal list returns FALLBACK_CATEGORY.

Configuration
-------------
Loaded from ``proxy.yml`` under ``reporting.signal_categories``.
Custom entries are *merged over* the defaults: unspecified defaults are preserved.
"""

from __future__ import annotations

import copy
import logging
from typing import Any

logger = logging.getLogger(__name__)


class SignalClassifier:
    """Maps sets of fired risk signal names to attack categories."""

    # Default mapping — mirrors PHASE_84.md §6.1 exactly.
    DEFAULT_SIGNAL_CATEGORIES: dict[str, dict[str, Any]] = {
        "spamhaus_drop":        {"category": "known_malicious_network",   "weight": 100},
        "spamhaus_edrop":       {"category": "known_malicious_network",   "weight": 100},
        "tor_exit":             {"category": "tor_exit_node",             "weight": 95},
        "beaconing_detected":   {"category": "c2_beaconing",              "weight": 90},
        "ja4_blacklist":        {"category": "malicious_tls_fingerprint", "weight": 85},
        "abuseipdb_score_high": {"category": "reported_abuse",            "weight": 70},
        "tls_version_old":      {"category": "obsolete_tls",              "weight": 60},
        "sni_missing":          {"category": "automation_tool",           "weight": 55},
        "sni_ip_literal":       {"category": "automation_tool",           "weight": 55},
        "datacenter":           {"category": "datacenter_scanner",        "weight": 50},
        "asn_datacenter":       {"category": "datacenter_scanner",        "weight": 50},
        "country_blacklist":    {"category": "geo_blocked",               "weight": 40},
    }
    FALLBACK_CATEGORY = "high_risk_score"

    def __init__(self, signal_categories: dict[str, dict[str, Any]] | None = None) -> None:
        """
        Args:
            signal_categories: Optional override dict.  Custom entries are merged
                *over* the defaults — all default entries not mentioned are kept.
                Pass ``{}`` to use pure defaults.  Pass ``None`` for the same.
        """
        merged = copy.deepcopy(self.DEFAULT_SIGNAL_CATEGORIES)
        if signal_categories:
            merged.update(signal_categories)
        self._categories: dict[str, dict[str, Any]] = merged

    @property
    def categories(self) -> dict[str, dict[str, Any]]:
        """Return a defensive copy of the active signal → {category, weight} mapping.

        The returned dict is a fresh copy; mutating it does not affect the classifier.
        """
        return {k: dict(v) for k, v in self._categories.items()}

    def classify(self, signals: list[str]) -> str:
        """Return the attack category for the given fired signals.

        Args:
            signals: List of RiskSignal names that fired for a connection.

        Returns:
            Category string (e.g. ``"known_malicious_network"``).
            Returns ``FALLBACK_CATEGORY`` when signals is empty or all unknown.
        """
        best_weight = -1
        best_category = self.FALLBACK_CATEGORY

        for signal in signals:
            entry = self._categories.get(signal)
            if entry is None:
                continue
            weight = entry["weight"]
            category = entry["category"]
            # Ties broken alphabetically by category name (ascending) so that
            # the result is deterministic regardless of iteration order.
            if weight > best_weight or (
                weight == best_weight and category < best_category
            ):
                best_weight = weight
                best_category = category

        return best_category

    def classify_batch(self, events: list[dict[str, Any]]) -> list[dict[str, Any]]:
        """Return a *new* list of events with a ``category`` field added to each.

        Does not mutate the input dicts.

        Args:
            events: List of connection event dicts.  Each event may have a
                ``signals`` key containing a list of signal name strings.
                Missing or ``None`` signals are treated as ``[]``.

        Returns:
            New list of dicts, each containing all original fields plus
            ``"category"``.
        """
        result = []
        for event in events:
            signals = event.get("signals") or []
            result.append({**event, "category": self.classify(signals)})
        return result
